src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,146 @@
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            firstNode = newNode;
+            lastNode = newNode;
+        } else {
+            newNode.prev = lastNode;
+            lastNode.next = newNode;
+            lastNode = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || size < index) {
+            throw new IndexOutOfBoundsException("Index out of range: " + index);
+        }
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, null);
+            newNode.next = firstNode;
+            firstNode.prev = newNode;
+            firstNode = newNode;
+            size++;
+        } else {
+            Node<T> oldNode = getNodebyIndex(index);
+            Node<T> newNode = new Node<>(null, value, null);
+            newNode.prev = oldNode.prev;
+            newNode.next = oldNode;
+            oldNode.prev.next = newNode;
+            oldNode.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> result = getNodebyIndex(index);
+        return result.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> result = getNodebyIndex(index);
+        T oldVal = result.value;
+        result.value = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (size == 0) {
+            throw new RuntimeException("Cannot remove from an empty list.");
+        }
+        return unlink(index).value;
     }
 
     @Override
     public boolean remove(T object) {
+        int i = 0;
+        for (Node<T> x = firstNode; x != null; x = x.next) {
+            if ((x.value == object || (x.value != null && x.value.equals(object)))) {
+                unlink(i);
+                return true;
+            }
+            i++;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private Node<T> getNodebyIndex(int index) {
+        checkIndex(index);
+        Node<T> node = firstNode;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
```

Comment:
as the Common mistakes says "Pay attention to the index you receive as an input, you can iterate only first or second half of the list, depending on the index value. That will boost your lists’ performance."